### PR TITLE
sql: support LATERAL joins

### LINKED ISSUE #3111: Support LATERAL in FROM clauses
Details [here](https://www.postgresql.org/docs/9.4/queries-table-expressions.html)

### LINKED ISSUE #3707: sql: decorrelation misplans non-cross-joined table functions
We currently plan all table functions as cross joins, even when they have been joined another way:

```
materialize=> create view t (a) as values (1), (3);
CREATE VIEW
materialize=> select * from t left join generate_series(t.a, 5) on generate_series > a;
 a | generate_series 
---+-----------------
 1 |               1
 1 |               2
 1 |               3
 1 |               4
 1 |               5
 3 |               3
 3 |               4
 3 |               5
(8 rows)
```

In postgres the (1, 1) and (3, 3) rows would be correctly omitted.

I tried to fix this (it's tightly related with supporting `LATERAL` derived tables), but it is mind-bindingly hard to do.

### PR DESCRIPTION
**Pull request by benesch**
~This is still pretty panicky on some of the Postgres test cases, but it mostly sorta works, so putting up in case anyone is curious.~

A LATERAL join allows the right-hand side of the join to access columns
defined on the left-hand side of the join. A simple example from the
PostgreSQL docs is the query

    SELECT * FROM foo, LATERAL (SELECT * FROM bar WHERE bar.id = foo.bar_id) ss

which is equivalent to:

    SELECT * FROM foo, bar WHERE bar.id = foo.bar_id;

The hope is that LATERAL joins will be useful for expressing "top-k
within a group", as in:

    SELECT * FROM
        (SELECT DISTINCT cat FROM foo) grp,
        JOIN LATERAL (SELECT * FROM foo WHERE foo.cat = grp.cat ORDER BY foo.val LIMIT $k)

Fixes #3111.
Fixes #3707.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/3713)

### CODE REVIEW

**Comment by benesch** [comment]
Ok, this is ready for a look. Might make sense to sit down together @frankmcsherry so I can walk you through the approach.

We pass all the PostgreSQL lateral join tests that I managed to port, save for two that cause "arrangement alarmingly absent" panics that I am hopeful are not my fault. Getting these tests to pass required fixing various unrelated bugs in the SQL planner, which is why there are so many commits in this stack.

The performance of many of these queries is abysmal, so we'll just have to warn folks to be really careful with these joins until our optimizer gets better (if ever; lateral joins are a great way to write n^2^2^2^2 queries).

**Comment by benesch** [comment]
Some debugging info from one of the panicky queries: https://gist.github.com/benesch/288674bae2fcd6376ceaa2714547ed03

File: test/sqllogictest/table_func.slt [hunk 1]
```diff
@@ -185,8 +194,18 @@ EXPLAIN PLAN FOR SELECT * FROM x, generate_series(1, a)
...
+| Get materialize.public.x (u5)
 | FlatMap generate_series(1, #0)
-| | demand = (#0..#2)
+| | demand = (#0, #2)
+
+%2 =
+| Join %0 %1 (= #0 #2)
+| | implementation = Differential %1 %0.(#0)
+| | demand = (#0, #1, #4)
+| Project (#0, #1, #4)
```
Comment (frankmcsherry) [inline test/sqllogictest/table_func.slt:208]: This is probably a beefy regression of a class we should sort out, somehow. I expect this falls under the general heading of "branch doesn't always need to join the left side again". We should at least decide who's job we want to make doing that, thoug!

Comment (frankmcsherry) [reply test/sqllogictest/table_func.slt:208]: Same for the subsequent `EXPLAIN` diffs; certainly going to be perf regressions there, pretty substantial to for something like json decoding (`FlatMap` is currently stateless; these joins will break that and have us keep the whole input collection around).

File: src/sql/src/plan/decorrelate.rs [hunk 2]
```diff
@@ -427,6 +454,32 @@ impl RelationExpr {
...
+    /// branch.
+    fn is_branch_safe(&self) -> bool {
+        let mut is_safe = true;
+        // Exhaustive match to prevent accidents when expression variants
+        // change.
+        self.visit(&mut |expr| match expr {
+            RelationExpr::Distinct { .. }
+            | RelationExpr::Reduce { .. }
+            | RelationExpr::Threshold { .. }
+            | RelationExpr::Negate { .. }
```
Comment (frankmcsherry) [inline src/sql/src/plan/decorrelate.rs:468]: I wasn't sure that these two need to be in here. Do you have a sense for if they need to be, or if they just weren't obviously safe?

Comment (benesch) [reply src/sql/src/plan/decorrelate.rs]: The latter! Anyway, reworked, see incoming comment.

**Comment by benesch** [comment]
Ok, I think this is in a place where this can merge. I adjusted the new branch planning to be very, very targeted; see here: https://github.com/MaterializeInc/materialize/pull/3713/commits/d37fe57a022b3fddd737f4cf1dd2cbc1937e7133#diff-ce359fcdca08c6161d027c3fd75d8b39R633-R690

With this new approach, there are (actually!) zero planning regressions—at least as far as our tests can tell.

**Comment by frankmcsherry** [comment]
Looks good to me!

**Pull request merged**